### DISCUSSION TITLE
How to add a column of computed fields and filter by that column

### DISCUSSION THREAD

**Discussion by bobmk2** [prisma/prisma#16718] [Q&A] [answered]
I would like to use Prisma to get filtered rows by reference to a field containing the computed results, but I could not figure out how to do this.

If I want these results, do I need to use a raw query?

## Table example

```prisma
model Rect {
  name String @db.VarChar(255)
  width Int
  height Int
}
```

```
> select * from sample;
+--------+-------+--------+
| name   | width | height |
+--------+-------+--------+
| Rect A |    10 |      5 |
| Rect B |     4 |      7 |
| Rect C |    12 |      3 |
+--------+-------+--------+
```

## Result I want

```
>> SELECT * FROM (SELECT *, (width * height) AS area from sample) as a  WHERE 20 <= a.area AND a.area <= 40;
+--------+-------+--------+------+
| name   | width | height | area |
+--------+-------+--------+------+
| Rect B |     4 |      7 |   28 |
| Rect C |    12 |      3 |   36 |
+--------+-------+--------+------+
```

**Comment by nurul3101** [comment, marked as answer]
Hello @bobmk2 👋 

You can use the newly released Prisma Client Extensions to add a computed column, here is an [docs example](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/result#re-use-a-computed-field-in-another-computed-field) which would be helpful in this case.

For your use case, you can achieve the area by extending Prisma Client like this:
```typescript
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

const xprisma = prisma.$extends({
  result: {
    rect: {
      area: {
        // the dependencies
        needs: { width: true, height: true },
        compute(rect) {
          // the computation logic
          return rect.width * rect.height;
        },
      },
    },
  },
});

async function main() {
  // Create React
  await prisma.rect.create({
    data: {
      height: 10,
      width: 20,
      name: 'rect1',
    },
  });
  const rect = await xprisma.rect.findFirst({});
  console.log(rect);
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
```
The response would contain the computed column:
<img width="539" alt="Screenshot 2022-12-15 at 6 38 10 PM" src="https://user-images.githubusercontent.com/13049676/207866971-98cfb8a3-b3dd-4ea4-935a-1f63e25494a5.png">

You can also use the Raw Query if extending Prisma Client is not an option.

**Comment by bobmk2** [comment]
Hi @nurul3101 

Thank you for your help and sorry for the delay in replying.

I can confirm that `area` can be taken with the sample code you provided. :tada:
This `$extends` is a great feature!

However, I would like to ask you one question.

I tried to add `area` with `where` to get more than 20, but I get the following error.
Do I need to set up anything additional to narrow down the calculated fields?

### Sample

```typescript
async function main() {
  // Create React
  await prisma.rect.create({
    data: {
      height: 10,
      width: 20,
      name: 'rect1',
    },
  });
  const rect = await xprisma.rect.findMany({
    // vvv add vvv
    where: {
      area: {
        gte: 20
      }
    }
   // ^^^^^^^^^^^
  });
  console.log(rect);
}
```

### Output

```
Unknown arg `area` in where.area for type RectWhereInput.
```

### Version

```
prisma: v4.8.0
```

**Reply by nurul3101** [reply]
Apologies for delay in getting back to you, this is because the computed field doesn't exist in the database so you cannot use it for filtering.

**Comment by pencilcheck** [comment]
I wonder if you can filter by computed fields. if not then it is a bummer

**Reply by nurul3101** [reply]
@pencilcheck As the computed field doesn't exist in the database, it can't be used for filtering records.

**Comment by ahmadkam72** [comment]
If you're getting an error stating that the column "your_calculated_field" does not exist, it could be due to the fact that calculated fields defined in your SELECT clause are not directly accessible in the WHERE clause in a standard SQL query.

To work around this limitation, you can use a common table expression (CTE) or a subquery to calculate the field and then filter based on the result

```
const result = await prisma.$queryRaw`
  SELECT *
  FROM (
    SELECT *, your_calculated_field + 10 AS modified_calculated_field
    FROM YourModel
  ) AS subquery
  WHERE modified_calculated_field = ${yourCondition};
`;
```
